File: server/src/main/com/broll/gainea/server/core/cards/impl/play/C_Panic.java
package com.broll.gainea.server.core.cards.impl.play;

import com.broll.gainea.server.core.cards.Card;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.objects.MapObject;
import com.broll.gainea.server.core.utils.UnitControl;
import com.google.common.collect.Lists;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class C_Panic extends Card {
    public C_Panic() {
        super(46, "Massenpanik", "Wählt ein Land mit mindestens einer Einheit. Alle Einheiten des gewählten Ortes werden auf angrenzende Orte verteilt.");
        setDrawChance(0.9f);
    }

    @Override
    public boolean isPlayable() {
        return true;
    }

    @Override
    protected void play() {
        List<Location> locations = game.getMap().getAllAreas().stream().filter(it -> !it.getInhabitants().isEmpty()).collect(Collectors.toList());
        if (!locations.isEmpty()) {
            Location source = selectHandler.selectLocation("Wählt einen Zielort für die Panik", locations);
            List<MapObject> inhabitants = new ArrayList<>(source.getInhabitants());
            List<Location> neighbours = new ArrayList<>(source.getConnectedLocations());
            Collections.shuffle(neighbours);
            int index = 0;
            for (MapObject object : inhabitants) {
                Location target = neighbours.get(index);
                UnitControl.move(game, Lists.newArrayList(object), target);
                index++;
                if (index >= neighbours.size()) {
                    index = 0;
                }
            }
        }
    }

}

File: server/src/main/com/broll/gainea/server/core/cards/impl/play/C_Panic.kt
package com.broll.gainea.server.core.cards.impl.play

import com.broll.gainea.server.core.cards.Card
import com.broll.gainea.server.core.cards.EffectType.DISRUPTION
import com.broll.gainea.server.core.utils.UnitControl.move

class C_Panic : Card(
    46,
    DISRUPTION,
    "Massenpanik",
    "Wählt ein Land mit mindestens einer Einheit. Alle Einheiten des gewählten Ortes werden auf angrenzende Orte verteilt."
) {
    init {
        drawChance = 0.9f
    }

    override val isPlayable: Boolean
        get() = true

    override fun play() {
        val locations = game.map.allAreas.filter { it.units.isNotEmpty() }
        if (locations.isNotEmpty()) {
            val source =
                selectHandler.selectLocation("Wählt einen Zielort für die Panik", locations)
            val neighbours = source.connectedLocations.toList().shuffled()
            var index = 0
            for (inhabitant in source.units.toList()) {
                val target = neighbours[index]
                game.move(inhabitant, target)
                index++
                if (index >= neighbours.size) {
                    index = 0
                }
            }
        }
    }
}
